qurator/eynollah/processor.py

```diff
@@ -1,68 +1,61 @@
+        self.models = None
+        if self.parameter['textline_light'] and not self.parameter['light_mode']:
+            raise ValueError("Error: You set parameter 'textline_light' to enable light textline detection but parameter 'light_mode' is not enabled")
 
-    def process(self):
-        LOG = getLogger('eynollah')
-        assert_file_grp_cardinality(self.input_file_grp, 1)
-        assert_file_grp_cardinality(self.output_file_grp, 1)
-        for n, input_file in enumerate(self.input_files):
-            page_id = input_file.pageId or input_file.ID
-            LOG.info("INPUT FILE %s (%d/%d) ", page_id, n + 1, len(self.input_files))
-            pcgts = page_from_file(self.workspace.download_file(input_file))
-            LOG.debug('width %s height %s', pcgts.get_Page().imageWidth, pcgts.get_Page().imageHeight)
-            self.add_metadata(pcgts)
-            page = pcgts.get_Page()
-            # XXX loses DPI information
-            # page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_filter='binarized')
-            image_filename = self.workspace.download_file(next(self.workspace.mets.find_files(local_filename=page.imageFilename))).local_filename
-            eynollah_kwargs = {
-                'dir_models': self.resolve_resource(self.parameter['models']),
-                'allow_enhancement': False,
-                'curved_line': self.parameter['curved_line'],
-                'full_layout': self.parameter['full_layout'],
-                'allow_scaling': self.parameter['allow_scaling'],
-                'headers_off': self.parameter['headers_off'],
-                'tables': self.parameter['tables'],
-                'override_dpi': self.parameter['dpi'],
-                'logger': LOG,
-                'pcgts': pcgts,
-                'image_filename': image_filename
-                }
-            Eynollah(**eynollah_kwargs).run()
-            file_id = make_file_id(input_file, self.output_file_grp)
-            pcgts.set_pcGtsId(file_id)
-            self.workspace.add_file(
-                ID=file_id,
-                file_grp=self.output_file_grp,
-                pageId=page_id,
-                mimetype=MIMETYPE_PAGE,
-                local_filename=join(self.output_file_grp, file_id) + '.xml',
-                content=to_xml(pcgts))
+    def process_page_pcgts(self, *input_pcgts: Optional[OcrdPage], page_id: Optional[str] = None) -> OcrdPageResult:
+        assert input_pcgts
+        assert input_pcgts[0]
+        assert self.parameter
+        pcgts = input_pcgts[0]
+        page = pcgts.get_Page()
+        # if not('://' in page.imageFilename):
+        #     image_filename = next(self.workspace.mets.find_files(local_filename=page.imageFilename)).local_filename
+        # else:
+        #     # could be a URL with file:// or truly remote
+        #     image_filename = self.workspace.download_file(next(self.workspace.mets.find_files(url=page.imageFilename))).local_filename
```

Comment:
```suggestion
```

This whole effort was to ensure we can pass a working local filename, as (was) needed by Eynollah. The approach by OCR-D is `Workspace.image_from_page` / `Workspace.image_from_segment` which will search for the right original _or_ derived image, download it if necessary and load it into memory.

I don't recall what the new behaviour of Eynollah is. If _both_ an image filename _and_ an image object are passed, who wins?

Assuming it's the memory object: this can be removed. (But then I wonder why we _still_ pass the image filename at all...)

Reply:
Well, currently we have

```python
            if image_pil:
                self._imgs = self._cache_images(image_pil=image_pil)
            else:
                self._imgs = self._cache_images(image_filename=image_filename)
[...]
     def _cache_images(self, image_filename=None, image_pil=None):
         ret = {}
         if image_filename:
             ret['img'] = cv2.imread(image_filename)
             self.dpi = check_dpi(image_filename)
         else:
             ret['img'] = pil2cv(image_pil)
             self.dpi = check_dpi(image_pil)
```

`image_filename` is (should) then only used passively, to generate filenames of plotted debug images as well as for PAGE serialization.

So I **think** `image_pil` should win but for now we need both. But as I said above, one of those things I would love to untangle in the refactoring.